internal/provider/resource_packer_channel.go

```diff
@@ -0,0 +1,381 @@
+	var channel packermodels.HashicorpCloudPackerChannel
+	for _, c := range resp.Channels {
+		if c.Slug == channelName {
+			channel = *c
+			break
+		}
+	}
+	if channel.ID == "" {
+		return diag.Errorf("Unable to find channel in bucket %s named %s.", bucketName, channelName)
+	}
+	return setPackerChannelResourceData(d, &channel)
+}
+
+func resourcePackerChannelCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	bucketName := d.Get("bucket_name").(string)
+	channelName := d.Get("name").(string)
+
+	client := meta.(*clients.Client)
+	loc := &sharedmodels.HashicorpCloudLocationLocation{
+		OrganizationID: client.Config.OrganizationID,
+		ProjectID:      client.Config.ProjectID,
+	}
+	if err := setLocationData(d, loc); err != nil {
+		return diag.FromErr(err)
+	}
+
+	channel, err := clients.CreateBucketChannel(ctx, client, loc, bucketName, channelName)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if channel == nil {
+		return diag.Errorf("Unable to create channel in bucket %s named %s.", bucketName, channelName)
+	}
+
+	_, ok := d.GetOk("iteration_assignment")
+	if !ok {
+		return setPackerChannelResourceData(d, channel)
+	}
+
+	return resourcePackerChannelUpdate(ctx, d, meta)
+}
+
+func resourcePackerChannelUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	bucketName := d.Get("bucket_name").(string)
+	channelName := d.Get("name").(string)
+
+	client := meta.(*clients.Client)
+	loc := &sharedmodels.HashicorpCloudLocationLocation{
+		OrganizationID: client.Config.OrganizationID,
+		ProjectID:      client.Config.ProjectID,
+	}
+
+	if err := setLocationData(d, loc); err != nil {
+		return diag.FromErr(err)
+	}
+
+	var assignment *clients.ChannelIterationAssignment
+	assignmentConfig, ok := d.GetOk("iteration_assignment")
+	if ok {
+		if config, ok := assignmentConfig.([]interface{})[0].(map[string]interface{}); ok {
+			assignment = expandIterationAssignmentConfig(config)
+		}
+	}
+	channel, err := clients.UpdateBucketChannel(ctx, client, loc, bucketName, channelName, assignment)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return setPackerChannelResourceData(d, channel)
+}
+
+func resourcePackerChannelDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	bucketName := d.Get("bucket_name").(string)
+	channelName := d.Get("name").(string)
+
+	client := meta.(*clients.Client)
+	loc := &sharedmodels.HashicorpCloudLocationLocation{
+		OrganizationID: client.Config.OrganizationID,
+		ProjectID:      client.Config.ProjectID,
+	}
+	if err := setLocationData(d, loc); err != nil {
+		return diag.FromErr(err)
+	}
+
+	_, err := clients.DeleteBucketChannel(ctx, client, loc, bucketName, channelName)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func resourcePackerChannelImport(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
```

Comment:
I'm not super familiar with the codebase here, but is there a reason why `meta` is an `interface{}` if we always expect it to be a `*clients.Client`?

Reply:
Good question. The interface is to fulfill the function signature expected by the underlying terraform plugin SDK: in this case, the [StateContextFunc](https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/schema/resource_importer.go#L50). The `meta` argument provides access to the provider state, but generally resources only need the client out of it.

Reply:
thanks @bcmdarroch I was struggling to find how to word this without going too deep. 🙇‍♂️